Adds script for global field importance for clusters

akashenfelter opened the pull request:
A script to determine which input fields are globally most important when differentiating between clusters. 

@jaor, would it be appropriate to put in a PR to add `merge-with` to the WhizzML standard library?

jaor commented on cluster-classification/script.whizzml:
```diff
@@ -0,0 +1,70 @@
+;; appended with the centroids the instances are part of.
+
+(define (label-by-cluster cl-id) 
+  (let (cl-ds (get-in (fetch cl-id) ["dataset"]) 
```
we're just releasing a new version of whizzml where maps and lists can be used as procedures, so that the idiomatic way of writing the above will become:

```
((fetch cl-id) "dataset")
```

it's fine if you prefer to send a PR afterwards modifying that, but you can also wait a day and fix it in this PR.

If we go with the current version, the currently idiomatic way would be using `get`:

```
(get (fetch cl-id) "dataset")
```

akashenfelter replied:
Waiting and fixing things up for whizzml 0.17 sounds like a good idea. I'll make all those changes and wait for the next version.

jaor commented on cluster-classification/script.whizzml:
```diff
@@ -0,0 +1,70 @@
+
+(define (label-by-cluster cl-id) 
+  (let (cl-ds (get-in (fetch cl-id) ["dataset"]) 
+        cl-fields (get-in (fetch cl-id) ["input_fields"]))
```
we are actually caching the result of calling `fetch`, but it's nonetheless a relatively heavy op:

```
(let (cl (fetch cl-id)
      cl-ds (cl "dataset")
      cl-fields (cl "fields"))
  ...)
```

akashenfelter replied:
Excellent, this is how to get multiple elements from a single fetch. I didn't like how I was doing it at all.

jaor commented on cluster-classification/script.whizzml:
```diff
@@ -0,0 +1,70 @@
+;; Given a list of pairs, creates the equivalent map
+(define (list-to-map nested-list)
+  (let (key-list (map head nested-list)
+        value-list (map (lambda (x) (nth x 1)) nested-list))
```
in the new-style whizzml the lambda body will become

```
(lambda (x) (x 1))
```

ashenfad replied:
Jumping in - but since these lists have just two values, you can drop the lambda fn altogether in the new WhizzML and just use `last`.

jaor replied:
yes, even nicer!

jaor commented on cluster-classification/script.whizzml:
```diff
@@ -0,0 +1,70 @@
+        subset-values (map (lambda (x) (fn (get map1 x) (get map2 x))) 
+                           subset-keys) 
+        subset (make-map subset-keys subset-values)) 
+    (merge map1 (merge map2 subset))))
```
This is a very nice implementation! Here's an alternative that i
suspect is more efficient, at the cost of some elegance:

```
(let (map2-ex (iterate (r map2 k (keys map2))
                (if (contains? map1 k)
                    (assoc r k (fn (map1 k) (map2 k)))
                    r)))
  (merge map1 map2-ex))
```

If you submit a PR to incorporate merge-with to the stdlib, we might
go for performance, but if you find your implementation clearer and a
couple measures with the REPL's ,time function don't show it's very
slow, i'll be fine with your implementation too.

akashenfelter replied:
I'll do some tests with the REPL's time function. I do think my version is more transparent, but of course I wouldn't want a noticeable hit to performance.

jaor commented on cluster-classification/script.whizzml:
```diff
@@ -0,0 +1,70 @@
+  (let (bc-id (label-by-cluster cl-id)
+        labeled (get-in (fetch bc-id) ["output_dataset_resource"])
+        _ (wait labeled)
+        cl-en (create-and-wait-ensemble {"dataset" labeled 
```
in the new version of whizzml, waits happen automatically for you and
create functions accept ids too, so you'll be able to write the above as:

```
(let (bc-id (label-by-cluster cl-id)
      labeled ((fetch bc-id) "output_dataset")
      cl-en (create-ensemble labeled {"randomize" true})
```

jaor commented on cluster-classification/script.whizzml:
```diff
@@ -0,0 +1,70 @@
+        imp-sum (reduce (lambda (x y) (merge-with + x y)) {} imp-map))
+    (delete labeled)
+    (delete bc-id)
+    (delete cl-en)
```
you can write that more succintly as:

```
(delete* [labeled bc-id cl-en])
```

jaor commented on cluster-classification/script.whizzml:
```diff
@@ -0,0 +1,70 @@
+    (delete bc-id)
+    (delete cl-en)
+    (map (lambda (x) (list (head x) (/ (nth x 1) len-models))) 
+         imp-sum)))
```
hmm, is this corret? `imp-sum` is a map, isn't it?

akashenfelter replied:
`imp-sum` is a map. Here I'm doing two things at once: converting `imp-sum` to a list of pairs (which is how importance is stored in the model object) and normalizing the summed importance value. Is it abusing `map` to use it to convert maps to lists? Should I pull this apart to make it more clear what's going on?

jaor replied:
hmmm, this is funny... the ability to map over maps (and their automatic
transformation to lists of pairs) is actually an accidental feature of
whizzml! i didn't implement it on purpose (and that's why it's
undocumented), and i'm not sure whether it'll work consistently.  but
since it's already working for you and you've found it useful, i'll just
write tests for it and document it in whizzml. thanks!

whizzmler replied:
although i'm having second thoughts on this... if we're going to allow
map over maps, it seems more natural to implement it so that the
function takes two arguments, key and value, so that you could write
your function:

```
 (map (lambda (k v) [k (/ v len-models)]) imp-sum)
```

and for multiple maps

```
 (map (lambda (k0 v0 k1 v1) ...) m0 m2)
```

and we can even mix maps and lists:

```
 (map (lambda (k0 v0 l k1 v1) ...) m0 ls m2)
```

whizzmler replied:
and, regardless of what we finally do with map over maps, yours right
now is easier to write and read like this (this already works):

```
(map (lambda (h t) [h (/ t len-models)]) (keys imp-sum) (values imp-sum))
```

since map can take multiple lists (and [x y] and (list x y) are equivalent).

ashenfad replied:
What about adding Clojure-like bindings in future versions of WhizzML?  So the lambda fn would still take a single var, but the binding gets the key and value:

``` clojure
(map (lambda ((k v)) ... ) some-hashmap)
```

jaor replied:
that's hard to implement, and the other solutions work good enough to
not completely justify that effort

jaor commented:
looks pretty good, just a few comments.  most of them are about adapting the code to whizzml 0.17, which will be released today or tomorrow... we can either wait for it, or merge this in old-style and then you submit a new PR to "modernize" it.

also, it'd be very nice if you could write a test (using bigmler as the driver) similar to those in `cross-validation` or `boruta`... they're essentially a bash script that uses bigmler to run the script and then check that the result make sense.

akashenfelter commented:
Will do! I'll add the tests using `cross-validation` and `boruta` as examples to guide me. Thanks!

akashenfelter commented:
Nits picked, rebased and squashed. 

I also made some changes to the output field; it now contains the field name, for much greater human readability. It was a little difficult because the batchcentroid field ids aren't the same as the dataset field ids. Is this something that could be changed?

The test script is coming in a future pull request, after I have a machine that can more easily run bash scripts. This pull request has been sitting so long, I thought it worth committing and closing even though it's not as nice as I'd like.

jaor commented on cluster-classification/script.whizzml:
```diff
@@ -0,0 +1,84 @@
+
+(define (label-by-cluster cl-id) 
+  (let (cl (fetch cl-id)
+        cl-ds (get cl "dataset")
```
we could remove all get and get-in...

jaor commented on cluster-classification/readme.md:
```diff
@@ -0,0 +1,11 @@
+# Global Field Importance for Clusters
```
perhaps we can also include this one in the top-level metadata.json as a component? and surely to the top-level readme.md in the examples list?

jaor commented:
looks pretty good: just a couple nits.

akashenfelter commented:
New nits picked! Bash shell comes to Windows 10 tomorrow, so that test script will be here very soon.

jaor commented:
looks good.  looking forward to that test script! merged, thanks a lot!
